app/app.go

```diff
@@ -158,20 +169,28 @@ func (app *MerkleEyesApp) doTx(tree merkle.Tree, tx []byte) abci.Result {
 
 func (app *MerkleEyesApp) Commit() abci.Result {
 
-	hash := app.state.Commit()
-
+	app.hash = app.state.Hash()
 	app.height++
+	fmt.Printf("height=%d,hash=%x\n", app.height, app.hash)
+
 	if app.db != nil {
-		app.db.Set(eyesStateKey, wire.BinaryBytes(MerkleEyesState{
-			Hash:   hash,
+		// This is in the batch with the Save, but not in the tree
+		tree := app.state.Append().(*iavl.IAVLTree)
```

Comment:
Atomicity. The app state should be written in the same batch write as the tree (otherwise if the process crashes in between, the persistent state won't match the saved tree). So I exposed BatchSet, which places the state in the same LevelDB batch that the tree will get into. To make that work I call Hash first, then Commit afterward.